rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,602 @@
+        :param AbstractTypestr type_: the type of the constant
+        :param value: the value of the constant
+        """
+        super().__init__()
+        assert isinstance(type_, AbstractType)
+        self.name = name
+        self.type = type_
+        self.value = value
+
+
+class Message:
+    """A structure containing constants and enums."""
+
+    __slots__ = ('structure', 'constants', 'enums')
+
+    def __init__(self, structure):
+        """
+        Constructor.
+
+        :param Structure structure: the structure of the message
+        """
+        super().__init__()
+        assert isinstance(structure, Structure)
+        self.structure = structure
+        self.constants = OrderedDict()
+        self.enums = []
+
+
+class Service:
+    """A namespaced type containing a request and response message."""
+
+    __slots__ = ('structure_type', 'request_message', 'response_message')
+
+    def __init__(self, type_, request, response):
+        """
+        Constructor.
+
+        :param NamespacedType type_: the namespaced type identifying the
+          service
+        :param Message request: the request message
+        :param Message response: the response message
+        """
+        super().__init__()
+
+        assert isinstance(type_, NamespacedType)
+        self.structure_type = type_
+
+        assert isinstance(request, Message)
+        assert request.structure.type.namespaces == type_.namespaces
+        assert request.structure.type.name == type_.name + \
+            SERVICE_REQUEST_MESSAGE_SUFFIX
+        self.request_message = request
+
+        assert isinstance(response, Message)
+        assert response.structure.type.namespaces == type_.namespaces
+        assert response.structure.type.name == type_.name + \
+            SERVICE_RESPONSE_MESSAGE_SUFFIX
+        self.response_message = response
+
+
+class Action:
+    """A namespaced type of an action including the derived types."""
+
+    __slots__ = (
+        'structure_type', 'goal_request', 'result_response', 'feedback',
+        'goal_service', 'result_service', 'feedback_message')
+
+    def __init__(self, type_, goal_request, result_response, feedback):
+        """
+        Constructor.
+
+        From the provide type the actually used services and message are
+        derived.
+
+        :param NamespacedType type_: the namespaced type identifying the action
+        :param Message goal_request: the goal request message
+        :param Message result_response: the result response message
+        :param Message feedback: the feedback message
+        """
+        super().__init__()
+
+        assert isinstance(type_, NamespacedType)
+        self.structure_type = type_
+
+        assert isinstance(goal_request, Message)
+        assert goal_request.structure.type.namespaces == type_.namespaces
+        assert goal_request.structure.type.name == type_.name + \
+            ACTION_GOAL_SERVICE_SUFFIX + SERVICE_REQUEST_MESSAGE_SUFFIX
```

Comment:
I think the suffix is needed to avoid potential collisions. The idea is that each type of interface lives in its own "namespace" without colliding with items of other types. E.g. a package can define a `Fibonacci.msg`, `Fibonacci.srv` and `Fibonacci.action` without the risk of collisions.

Reply:
I thought `ACTION_GOAL_SERVICE_SUFFIX` by containing an underscore avoided the namespace collision already.

Reply:
Can you describe the exact change you propose.

Do you want the user defined goal / result types to not have a request / response suffix?